### ISSUE TITLE
monitored params for WAIC fail for state-space model

### ISSUE THREAD

**Issue by danielturek** [nimble-dev/nimble#799]
Reproducible example below.

Problem has to do with what WAIC thinks it does (or doesn't) need monitored for WAIC, which it gets wrong in this case, I think because of the deterministic declarations for `x[1,1]` and `y[1,1]`.

```
set.seed(0)
k <- 10
n <- 75
phi <- .9
p <- 0.5
y <- array(NA, c(n,k))

for(i in 1:n) {
    y[i,1] <- 1
    x <- 1
    for(t in 2:k) {
        x <- rbinom(1, prob=x*phi, size=1)
        y[i,t] <- rbinom(1, prob=p*x, size=1)
    }
}

library(nimble)

(N <- dim(y)[1])
(T <- dim(y)[2])

code <- nimbleCode({
    p   ~ dbeta(1,1)
    phi ~ dbeta(1,1)
    for(i in 1:N) {
        x[i,1] <- 1
        y[i,1] <- 1
        for(t in 2:T) {
            x[i,t] ~ dbern(phi * x[i,t-1])
            y[i,t] ~ dbern(p * x[i,t])
        }
    }
})

constants <- list(N = N, T = T)
data <- list(y = y)
inits <- list(x = (1^y), p = 0.5, phi = 0.5)
Rmodel <- nimbleModel(code, constants, data, inits)
Rmodel$calculate()    ## -935.7487

conf <- configureMCMC(Rmodel)
Rmcmc <- buildMCMC(conf, enableWAIC = TRUE)
compiledList <- compileNimble(list(model=Rmodel, mcmc=Rmcmc))
Cmodel <- compiledList$model; Cmcmc <- compiledList$mcmc

Cmodel$calculate()   ## -935.7487

set.seed(0)
out <- runMCMC(Cmcmc, 10000, WAIC = TRUE)

out$WAIC
## [1] NaN

Cmodel$calculate()
## -Inf
```

**Comment by danielturek** [comment]
I should also have added, that if `x` is monitored also, then everything works fine.

e.g., WAIC's determination of "what needs to be monitored for WAIC to work" should also include `x` here (I believe)

**Comment by NLMichaud** [comment]
Thanks a bunch for finding this and providing such a useful example Daniel!  You're correct that the issue here arises from calculating the WAIC after not monitoring `x` during the MCMC.  However, I'm not sure that the deterministic nature of `x` is the cause of the miscalculation.  

Within the WAIC algorithm, we simulate values for un-monitored stochastic nodes that are downstream of the monitored parameters (here, monitored parameters are `phi` and `p`), and "upstream" of data nodes.  When `x` is not monitored, this means that the algorithm is simulating values for `x` conditioned on the sampled values for `phi` and `p`.  This model has the interesting feature that if an `x[i,j]` value is simulated to be 0, when `y[i,j]` is 1, then log( p(y[i,j] | x[i,j]) ) (as determined by `model$calculate(y[i,j])`) will be `-Inf`. However, this causes a problem, as the WAIC algorithm (as given in equation 12 of [this paper](http://www.stat.columbia.edu/~gelman/research/published/waic_understand3.pdf)) requires calculation of the posterior variance of the log predictive density, which involves simulating an `x[i,j]` value from each posterior value of `phi` and `p`, and taking the variance of log( p(y[i,j] | x[i,j]) ) over all such simulated `x[i,j]`.  When any such `log( p(y[i,j] | x[i,j]) ) = -Inf`, this variance becomes `NaN`.

It seems as though the WAIC calculation is not designed to handle cases when e.g. ` log( p(y[i,j] | x[i,j]) ) = -Inf`.  These cases seem like they can can arise in models when intermediate stochastic nodes are not monitored, and simulated values for these nodes disagree entirely with the observed data.  

It's not clear to me that there's a good fix for this, as in some sense the `NaN` value is the WIAC algo performing according to its definition.  One option would be to emit a warning if the WAIC returns an `NaN` value to alert the user that:
a) Some sets of monitored nodes, even if technically valid for calculating a WAIC value, can lead to `NaN` WAICs such as above
b) Even if all stochastic nodes are monitored, if inserting any set of posterior values for these nodes into the model ever leads to a `model$calculate()` value of `-Inf` (unclear when or if this would be the case but may be possible?), then again we would expect an `NaN` WAIC value

I'm open to any other suggestions for how to address this behavior, and please let me know if the explanation above was unclear as well.  Thanks!

**Comment by danielturek** [comment]
@NLMichaud Thanks for the full report.  It sounds like there are some questions here about how we want WAIC to behave, although it does sound like the algorithm is operating as intended.

Just one more thing to point out, which is still disturbing to me.  Even if WAIC shouldn't work in this case, e.g. should return `NaN` for the reasons you explained.  What's troubling to me is that after calling `Cmcmc$calculateWAIC()`, it leaves the `Cmodel` object in a **different** state than originally.  Specifically, the "calculation" of WAIC overwrites the values of `x` in the model, presumably with the newly simulated values, which are no longer valid, and thus `Cmodel$calculate()` now returns -Inf.

One easy & quick fix for this, would be instead of the WAIC algorithm just saving part of the model state as `currentVals <- values(model, sampledNodes)`, then restoring these values at the end with:  `values(model, sampledNodes) <<- currentVals`, changing these two lines to:

```
currentVals <- values(model)
## WAIC calculation here
values(model) <<- currentVals
model$calculate()
```

in a very defensive manner, so that regardless of what happens in the WAIC calculation (correct, `NaN`, or whatever!), the model state is guaranteed to be returned to the same, original, presumably valid state, that is was in **before** the WAIC "calculation".

What do you think @NLMichaud?  I'm happy to make this change, if you agree.

**Comment by danielturek** [comment]
Great, will do, Nick.  Thanks for getting back.

**Comment by danielturek** [comment]
@perrydv @NLMichaud @paciorek I've resolved the issue of the model changing, using Chris's suggestion of "saving" the model values and also the logProb values, rather than using a model$calculate().   This was fixed in PR #808.

I also added a warning message, when WAIC is calculated as `NaN`, indicating that latent states may also need to be monitored.
